### PR TITLE
Migrate RecursionIT, and materialisation tests from TransactionIT, to BDD

### PR DESCRIPTION
**Pull request by alexjpwalker**
## What is the goal of this PR?

To migrate RecursionIT, and materialisation tests from TransactionIT, from grakn to BDD.

## What are the changes implemented in this PR?

Added new tests in `recursion.feature` and `insert.feature` which were migrated from grakn.

Also, moved the Chained Inference tests from `relation-inference` to `recursion`.

(Note: There are a couple of unrelated tests lumped in (to other files), in response to conversations we've been having recently)

### CODE REVIEW

File: behaviour/graql/reasoner/recursion.feature [hunk 1]
```diff
@@ -0,0 +1,2025 @@
...
+        {$nameX == 'f';$nameY == 'k';};
+      get $x, $y;
+      """
+#    Then materialised and reasoned keyspaces are the same size
+
+
+  # TODO: re-enable all steps when resolvable (currently takes too long) (#75)
+  Scenario: dual linear transitivity matrix test
+
+    test 6.1 from Cao p 71
```
Comment (kasper-piskorski) [inline behaviour/graql/reasoner/recursion.feature:1312]: Just to give the full reference. It's Cao - Methods for evaluating queries toHorn knowledge bases in first-order logic

File: behaviour/graql/reasoner/recursion.feature [hunk 2]
```diff
@@ -0,0 +1,2025 @@
...
+      """
+      match $y isa b-entity; get;
+      """
+#    Then materialised and reasoned keyspaces are the same size
+
+
+  # TODO: re-enable all steps when resolvable (currently takes too long) (#75)
+  Scenario: Nguyen's test
+
+    test3 from Nguyen (similar to test 6.5 from Cao):
```
Comment (kasper-piskorski) [inline behaviour/graql/reasoner/recursion.feature:1693]: Madalinska-Bugaj, Nguyen - A Generalized QSQR Evaluation Method for Horn Knowledge Bases

File: behaviour/graql/reasoner/recursion.feature [hunk 3]
```diff
@@ -0,0 +1,2025 @@
...
+      """
+      match $x has index 'a2'; get;
+      """
+#    Then materialised and reasoned keyspaces are the same size
+
+
+  # TODO: re-enable all steps when resolvable (currently takes too long) (#75)
+  Scenario: given a directed graph, all pairs of vertices (x,y) such that y is reachable from x can be found
+
+    test 5.2 from Green
```
Comment (kasper-piskorski) [inline behaviour/graql/reasoner/recursion.feature:929]: Green - Datalog and Recursive Query Processing

**Comment by alexjpwalker** [comment]
> added references
Thanks! I'll update the tests with the full names of these source materials.

File: behaviour/graql/reasoner/recursion.feature (hunk 1 above)
Comment (alexjpwalker) [reply behaviour/graql/reasoner/recursion.feature]: Citation added

File: behaviour/graql/reasoner/recursion.feature (hunk 2 above)
Comment (alexjpwalker) [reply behaviour/graql/reasoner/recursion.feature]: Citation added

File: behaviour/graql/reasoner/recursion.feature (hunk 3 above)
Comment (alexjpwalker) [reply behaviour/graql/reasoner/recursion.feature]: Citation added

File: behaviour/graql/reasoner/recursion.feature [hunk 4]
```diff
@@ -0,0 +1,2026 @@
...
+      get;
+      """
+#    Then all answers are correct in reasoned keyspace
+    Then answer size in reasoned keyspace is: 20
+    Then answer set is equivalent for graql query
+      """
+      match
+        $Y isa person, has name $nameY;
+        $X isa person, has name $nameX;
+        {$nameX == 'a';$nameY == 'aa';} or
```
[the remark this answers is not in the document]
Comment (alexjpwalker) [reply behaviour/graql/reasoner/recursion.feature:641]: You'd have to ask whoever wrote the test that it was migrated from!
In general these "explicit" queries (that verify that the answer set actually contains the expected answers) divide up the various disjunction blocks into sensible line lengths and consistently have the word 'or' at the end of the line.
As for the X/Y I guess we can just swap them.

[the remark this answers is not in the document]
Comment (alexjpwalker) [reply behaviour/graql/reasoner/recursion.feature:641]: Updated to swap the first two lines around.

**Pull request merged**